src/charm.py

```diff
@@ -379,25 +443,87 @@ def proceed_on_broken_event(self, event) -> bool:
 
         return True
 
-    def get_mongos_host(self) -> str:
-        """Returns the host for mongos as a str.
+    def get_units(self) -> List[Unit]:
+        units = [self.unit]
+        units.extend(self.peers_units)
+        return units
 
-        The host for mongos can be either the Unix Domain Socket or an IP address depending on how
-        the client wishes to connect to mongos (inside Juju or outside).
+    def get_mongos_hosts_for_client(self) -> Set:
+        """Returns the hosts for mongos as a str.
+
+        The host for mongos can be either the K8s pod name or an IP address depending on how
+        the app has been configured.
+        """
+        if self.is_external_client:
+            return self.get_ext_mongos_hosts()
+
+        return self.get_k8s_mongos_hosts()
+
+    def get_k8s_mongos_hosts(self) -> Set:
+        """Returns the K8s hosts for mongos"""
+        hosts = set()
+        for unit in self.get_units():
+            hosts.add(self.get_k8s_mongos_host(unit))
+
+        return hosts
+
+    def get_ext_mongos_host(self, unit: Unit, incl_port=True) -> str | None:
+        """Returns the ext hosts for mongos on the provided unit."""
+        if not self.is_external_client:
+            return None
+
+        try:
+            unit_ip = self.node_port_manager.get_node_ip(unit.name)
+            if not incl_port:
+                return unit_ip
+
+            unit_port = self.node_port_manager.get_node_port(
+                port_to_match=Config.MONGOS_PORT, unit_name=unit.name
+            )
+            if unit_ip and unit_port:
+                return "{unit_ip}:{unit_port}"
+            else:
+                raise NoExternalHostError(f"No external host for unit {unit.name}")
+        except (
+            NoExternalHostError,
+            FailedToFindNodePortError,
+            FailedToFindServiceError,
+        ) as e:
+            raise FailedToGetHostsError(
+                "Failed to retrieve external hosts due to %s", e
+            )
```

Comment:
**todo:** I **strongly** recommend you `sort()` your hosts. 

Off the top of my head, I see that you use these hosts in the `mongos_config`. Not sure what you do with this, but let's assume you use them in some env-var during service runtime:
```bash
HOSTS=host1,host2,host3 start-mongos.sh
```

Pebble replan looks for new runtime args, so if you now get:

```bash
HOSTS=host3,host1,host2 start-mongos.sh
```

you're going to restart without needing to.

Reply:
fortunately we do not provide hosts to pebble plan! But I suppose we do supply them to clients, I suppose we could lead to unnecessary relation-changed events.

Reply:
I didn't look deeply, but if you do:
```python
hosts = set(["1", "2", "3"])
relation.data[some.app].update({"hosts": ",".join(hosts)})
```

You'll get a random order, and have extraneous relation-changed events.

Reply:
I see , thank you very much @marcoppenheimer ! I think we ought to handle this in where we update hosts (the provider lib) which is on the mongodb-vm operator GitHub, I [added an issue for it](https://github.com/canonical/mongodb-operator/issues/491)